### PR TITLE
feat(qwen36): routed experts from a qpack container through bounded Metal slots

### PR DESCRIPTION
**Pull request by Avicennasis**
## What this adds

The Metal slot-pool half of the original series: `qwen36` runs a Swiftlet-format qpack container's routed experts through #1290's affine Metal path from a bounded, refillable slot pool. Dense weights and norms come from the loader PR this is stacked on (#1343).

Four commits of our own:

1. `feat(metal)`: whole-slot Metal buffers with offset-addressed affine dispatch (`coli_metal_slot_register` / `coli_metal_matmul_affine_slot`): a slot is wrapped zero-copy once (page-aligned only; copy-wrapping would silently detach the GPU from refills), projections addressed by `setBuffer:offset:` with bounds/alignment checks and a stale-descriptor tripwire. Host-pointer path and fmt=4 untouched. New cases in `tests/test_backend_metal.mm`.
2. `feat(qwen36)`: routed experts through bounded slots (`qwen36_qpack.{c,h}`): per-expert blobs read with one `pread` into an LRU pool (default 96, `QWEN36_QPACK_SLOTS=<n>`), per-slot generation counters checked at dispatch so a stale reference is refused rather than served; `moe()` gains the qpack branch ahead of the CUDA tier; `QWEN36_QPACK=<dir>` in `main` with container-vs-config refusal. CPU gate `tests/test_qwen36_qpack.c` (portable affine reference on a synthesized container; runs on every platform, no GPU).
3. `test(qwen36)`: Apple parity gates (`make qwen36-metal-test`, fixture-driven, loud skip marker when no fixture): one-layer MoE and prefill/decode logits vs the CPU reference under forced eviction, stale-reference refusal, resident-buffer accounting.
4. `feat(qwen36)`: end-of-run `[qpack] projections: metal=… cpu=… | slots=… fills=… evictions=…` counters, so a CPU fallback cannot masquerade as GPU parity.

Own diff: 9 files, +1775/-41 on top of the loader PR.

## Stacked on

#1343 (dense weights, norms, snap-view tool), which is itself stacked on #1290, #1319 and #1320. All of their commits show in this diff until they land: the first 10 commits here are the loader PR's branch verbatim.

Rebased onto `dev` at `f58a2679` on 2026-09-04. Two conflicts, both in commit 2:

- `c/Makefile`, the `qwen36$(EXE)` rule: kept dev's `cli_args.h` dependency and added `qwen36_qpack.h`, `$(QWEN36_QPACK_SRC)`, `$(METAL_OBJ)`. The test-rule block merged on its own; dev's new rules and ours are all present.
- `c/qwen36.c`, the CUDA tier init: #1334's `expert_is_int4` probe and 8-argument `qt_init` are kept as-is, with our `!qq_active() &&` guard in front of the call. The tier still learns int8 from the on-disk size, and is still skipped when the container owns the routed experts.

## Validation

Linux (x86-64, gcc), this branch at `121c663a`:

- `make -C c check`: exit 0. `test-c` ran 113 gate binaries (the loader PR's 112 plus `test_qwen36_qpack`), none failed; `test-python` 739 tests OK, 35 skipped.
- `make -C c test-asan`: exit 0, the same 113 binaries clean under ASan + UBSan.

Not re-run after the rebase (no Apple machine in this pass): the Metal side. CI's macOS lane compiles `tests/test_backend_metal.mm` through `make metal-test` (GPU submissions are skipped on the hosted runner's paravirtual device); `make qwen36 METAL=1` and the fixture-driven `make qwen36-metal-test` are not in CI and need a real Apple machine.

Pre-rebase Apple results, from this PR's original description (M4 Max, macOS 26.6.2, series head `926f19be`): `metal-test` unchanged-green; qpack gate MoE one-layer `nerr=4.86e-07 metal=18/18 cpu-fallback=0`, logits prefill `3.58e-07` / decode `5.85e-07`, `288/288` Metal dispatches under live eviction (`evictions=136`), stale ref refused. Real model `Leonickson/Qwen3.6-35B-A3B-qpack` (18 GB): `The capital of France is` → ` Paris, a city renowned for its iconic landmarks such as the Eiffel Tower,`, byte-identical to Swiftlet's output from the same bytes over 64 greedy tokens; `[qpack] projections: metal=19200 cpu=0 … fallbacks=0`, 11.3 GB peak RSS.

## Run recipe

```sh
make -C c qwen36 METAL=1          # without METAL=1 the qpack path is compiled out
python3 c/tools/make_qwen36_qpack_snap.py --container /path/to/model.qpack --out /path/to/snap
SNAP=/path/to/snap QWEN36_QPACK=/path/to/model.qpack N_NEW=16 ./qwen36 16 4 prompt.txt
```

## Scope

Attention, DeltaNet, dense projections, router, and the shared expert run on CPU in this series; only routed experts go to Metal. Async expert blocks and the dense sublayers on Metal are follow-ups.

### CODE REVIEW

**Comment by JustVugg** [comment]
Status from our side, and one thing blocking it that is yours rather than ours.

**It is conflicting with `dev` and has no CI runs at all** — GitHub will not schedule workflows for a PR it cannot merge, so this one is invisible to the gates while the rest of your series has been unblocked. I approved the pending runs on #1319, #1320, #1315 and #1316 just now; this one I could not, because there is nothing to approve.

A rebase onto current `dev` is what unsticks it. Worth knowing before you do: `dev` moved a fair bit today — #1334 (qwen36 VRAM tier learns int8), #1335 (Kimi DSA indexer gated off), #1336 (mid-turn CANCEL, new `serve_poll.h`), plus a new `cli_args.h` and several test rules in `c/Makefile`. The Makefile test-rule block is the likeliest place you will collide; keeping both sides is almost always the right resolution there, as those rules are what make a test a gate.

On the series itself, so the rebase is worth your time: the structure you describe is the right one. Commits 1 and 2 are already filed separately as #1319 and #1320 and will drop out when they land — that split is appreciated, it makes the reviewable part smaller. The remaining question for the review is scope: **+5398 lines across 25 files** stacked on #1290 is a lot to take in one piece, and #1290 itself is still under review. If the qpack loader path and the Metal slot pool can stand as separate PRs, each gets a real review instead of a skim; if they genuinely cannot be split, say so and it gets reviewed as one.

No rush from us — the queue is long and this is being tracked. But without the rebase it cannot even be tested, and that would be a waste of the work.

**Comment by Avicennasis** [comment]
Rebased onto `dev` at `f58a2679` (the #1336 merge) and split as you suggested.

**Split.** The loader path is now its own PR, #1343 (`feat/qwen36-qpack-loader`): #1290's four commits, #1319, #1320, then the four loader commits (dequant oracle, dense weights from the container, norm unshift, snap-view tool); own diff 8 files, +1221/-16. This PR keeps the Metal slot pool (whole-slot buffers with offset dispatch, routed experts through bounded slots, the Apple parity gates, the end-of-run counters; own diff 9 files, +1775/-41) and is stacked on the loader PR, so its 10 commits appear here until it lands. The halves do separate cleanly: the loader half has no Metal code, and the slot-pool half only consumes it. Title and body here are rewritten to describe the slot-pool half only.

**Conflicts.** Two, both in `feat(qwen36): routed experts from a qpack container through bounded slots`:

- `c/Makefile`, the `qwen36$(EXE)` rule: kept dev's `cli_args.h` dependency and added `qwen36_qpack.h`, `$(QWEN36_QPACK_SRC)`, `$(METAL_OBJ)`. The test-rule block itself merged automatically; dev's new rules (`test_qwen36_tier_int8`, `test_serve_poll`, `test_cli_args`) and ours are all present, so both sides' tests remain gates.
- `c/qwen36.c`, the CUDA tier init where #1334 added the `expert_is_int4` probe and the 8-argument `qt_init`: kept #1334's probe and signature as-is and put our `!qq_active() &&` guard in front of the call, so the tier still learns int8 from the on-disk size and is still skipped when the qpack container owns the routed experts.

No other hunk changed. `git range-diff` against the old series shows the other 13 commits patch-identical; #1319/#1320 were replayed from their current heads (`cfd51066`, `f01621ff`), which have the same patch-ids as the copies that were in this branch.

**Linux gates** (x86-64, gcc):

- Loader branch `05868e5b`: `make -C c check` exit 0 — `test-c` 112 gate binaries, none failed; `test-python` 739 OK, 35 skipped. `make -C c test-asan`: exit 0, the same 112 binaries clean under ASan + UBSan.
- This branch `121c663a`: `make -C c check` exit 0 — `test-c` 113 gate binaries (adds `test_qwen36_qpack`), none failed; `test-python` 739 OK, 35 skipped. `make -C c test-asan` exit 0, 113 binaries clean under ASan + UBSan.

**Not run here:** anything Metal. No Apple machine in this pass, so `make metal-test`, `make qwen36 METAL=1` and the fixture-driven `make qwen36-metal-test` were not re-run after the rebase. CI's macOS lane will compile `test_backend_metal.mm` through `metal-test`; the qwen36 `METAL=1` build and the parity gate are outside CI. The M4 Max figures in the body are from the pre-rebase series and are labelled as such.

`feat/qwen36-qpack-engine` force-pushed from `926f19be` to `121c663a`.

**Comment by Avicennasis** [comment]
Apple/Metal gates after the rebase, run today on two machines from fresh `/tmp` clones of the same bundle (`git checkout --detach` by SHA and `git clean -fdxq` before every target; Command Line Tools only, Apple clang 21.0.0, GNU Make 3.81, libomp from `/opt/homebrew/opt/libomp`, macOS 26.6.2 on both).

**M4 Max** (MacBook Pro Mac16,5):

| target | `dev` `f58a2679` | loader `05868e5b` (#1343) | this PR `121c663a` |
|---|---|---|---|
| `make -C c metal-test` | rc 0, 93 ok / 0 mismatch | rc 0, 99 ok / 0 mismatch | rc 0, 102 ok / 0 mismatch |
| `make -C c gemm-test` | not run | rc 0, 12/12, `GEMM large-batch: all clean` | rc 0, 12/12, `GEMM large-batch: all clean` |
| `make -C c qwen36 METAL=1` | not run | rc 0 | rc 0 |
| `make -C c {colibri,inkling,kimi_k3} METAL=1` | not run | rc 0 ×3 | rc 0 ×3 |
| `make -C c qwen36-metal-test` | no rule | no rule | rc 0; without the fixture `SKIPPED: QWEN36_QPACK_FIXTURE not set -- qpack parity gates need a Swiftlet fixture (swiftlet-repack of tiny-model-q4)`; with `QWEN36_QPACK_FIXTURE=<swiftlet-repack of tiny-model-q4>`: `qwen36 qpack metal parity: ok` |

Fixture parity at `121c663a`: moe one-layer nerr 4.86e-07 (metal 18/18, cpu-fallback 0); logits prefill 3.58e-07, decode 5.85e-07; routed dispatch metal 288/288; stale slot ref refused with zero dispatch and re-acquired on Metal 3/3; pool churn evictions=136 fills=141 pool=5; 5 resident slot buffers / 81920 bytes; affine handles released.

**M1 Max** (MacBookPro18,2), same three refs: `gemm-test`, every `METAL=1` build and `qwen36-metal-test` (identical parity numbers) pass as above, but `metal-test` exits 2 at all three refs — `dev` base 92 ok / 1 mismatch, loader 98 / 1, this PR 101 / 1 — always the same single line: `fp8 LUT exactness (256/256 codes via GPU kernel) 96/256 mismatches  *** MISMATCH`. That case (`run_fp8_lut`, on `dev` since `6ca61a1b`) fails at `dev` `f58a2679` without either PR and passes on the M4 Max at all three refs (`0/256 mismatches`), so it is pre-existing and platform-specific, not something these PRs introduce. The counts differ by exactly the new cases — +6 for the loader (`affine Q4|Q8 / f32|f16|bf16 batched`), +3 for the slot pool (`affine slot ... offset+refill`) — and all nine pass on both machines.

**Comment by JustVugg** [comment]
Hold this one until #1343 is rebased, then stack it back on top. Its file set is a strict superset of #1343's, so rebasing both in parallel would mean resolving the same `c/qwen36.c` conflicts twice.

Same churn underneath: `qwen36.c` 18 commits, `Makefile` 39, `st.h` 3, `backend_metal.mm` 3. Your own new files (`qwen36_qpack.{c,h}`) cannot conflict.

The Apple and Metal gate results you posted from two machines are what makes this reviewable at all, so thank you for those.

**Comment by Avicennasis** [comment]
Re-stacked on #1343's new head `f0733efa` (`dev` `a90bed95`): `feat/qwen36-qpack-engine` `121c663a` -> `8c8d890b`, the six engine commits on top of the loader's nine, 15 ahead / zero behind `dev`, loader head an ancestor. `git range-diff`: `dcf68533`, `77de6f19`, `c5c9d7bc`, `8f60fa21` identical; `ca92a5d5` differs only by the integration below; the one-line `backend_metal.h` prerequisite was re-applied to `dev`'s current line. Conflicts, all in `ca92a5d5`:
- `c/Makefile` `qwen36$(EXE)`: union of `dev`'s prerequisites (`expert_ffn.h omp_tune.h kv_prefix.h`) and the `METAL=1`-gated `QWEN36_QPACK_SRC` block, `qwen36_qpack.h`, `$(QWEN36_QPACK_SRC)`, `$(METAL_OBJ)`; the `tests/test_qwen36_qpack` recipe sits beside `dev`'s widened `test_qwen36_json_escape` line.
- `c/qwen36.c` include block: `dev`'s `expert_ffn.h` plus `qwen36_qpack.h`.
- `c/qwen36.c` `moe()`: `dev`'s new `expert_ffn.h` kernel (`use_xf`, `275a1ee0`) is a third routed-expert source, so it is gated exactly like the CUDA tier -- `use_xf = !use_qq && !use_qt && xf_mode(m)`, per-token chain `use_qq -> use_xf -> use_qt -> CPU`. Without that gate `moe_xf_run()` would run the routed experts a second time after the loop on a qpack run. Shared expert and combine stay on the CPU path as before.
Gates at `8c8d890b`. Linux x86-64 (gcc 13.3.0, Python 3.12.3): `make -C c colibri qwen36` and the five qpack/affine test binaries build; `tools/run_tests.py` 5/5 ok (`qwen36 qpack CPU gate: ok`); `tests.test_makefile_deps` 2 OK; `git diff --check` clean.
Apple M4 Max (macOS 26.6.2, Apple clang 21.0.0), fresh clone, checkout by SHA and `git clean -fdxq` per target: `make -C c qwen36 METAL=1` rc 0 (61 `qq_`/`coli_metal` symbols); `make -C c metal-test` rc 0, 106 ok / 0 mismatch, the three `affine slot ... offset+refill` cases at nerr 0; `make -C c qwen36-metal-test` rc 0 SKIPPED without a fixture, and with `QWEN36_QPACK_FIXTURE` set to the Swiftlet-generated tiny-model-q4 container `qwen36 qpack metal parity: ok` -- moe one-layer nerr 4.86e-07 (Metal 18/18, CPU fallback 0), logits prefill 3.58e-07 / decode 5.85e-07, routed dispatch 288/288, stale ref refused and re-acquired 3/3, 136 evictions / 141 fills in a five-slot pool, 5 buffers / 81,920 bytes, affine handles released.

**Comment by JustVugg** [comment]
Stays stacked on #1343; see the note there about `json.h` (reuse #1605's `json_parse_checked` instead of a second strict mode). Next release, together with #1343 and #1315.
